gateway/blocks_gateway.go

```diff
@@ -0,0 +1,484 @@
+		logical roots that represent each segment of the human-readable content
+		path:
+		Given contentPath = /ipns/en.wikipedia-on-ipfs.org/wiki/Block_of_Wikipedia_in_Turkey
+		rootCidList is a generated by doing `ipfs resolve -r` on each sub path:
+			/ipns/en.wikipedia-on-ipfs.org → bafybeiaysi4s6lnjev27ln5icwm6tueaw2vdykrtjkwiphwekaywqhcjze
+			/ipns/en.wikipedia-on-ipfs.org/wiki/ → bafybeihn2f7lhumh4grizksi2fl233cyszqadkn424ptjajfenykpsaiw4
+			/ipns/en.wikipedia-on-ipfs.org/wiki/Block_of_Wikipedia_in_Turkey → bafkreibn6euazfvoghepcm4efzqx5l3hieof2frhp254hio5y7n3hv5rma
+		The result is an ordered array of values:
+			X-Ipfs-Roots: bafybeiaysi4s6lnjev27ln5icwm6tueaw2vdykrtjkwiphwekaywqhcjze,bafybeihn2f7lhumh4grizksi2fl233cyszqadkn424ptjajfenykpsaiw4,bafkreibn6euazfvoghepcm4efzqx5l3hieof2frhp254hio5y7n3hv5rma
+		Note that while the top one will change every time any article is changed,
+		the last root (responsible for specific article) may not change at all.
+	*/
+	var sp strings.Builder
+	var pathRoots []cid.Cid
+	contentPathStr := contentPath.String()
+	pathSegments := strings.Split(contentPathStr[6:], "/")
+	sp.WriteString(contentPathStr[:5]) // /ipfs or /ipns
+	var lastPath ifacepath.Resolved
+	for _, root := range pathSegments {
+		if root == "" {
+			continue
+		}
+		sp.WriteString("/")
+		sp.WriteString(root)
+		resolvedSubPath, err := api.resolvePath(ctx, ifacepath.New(sp.String()))
+		if err != nil {
+			// TODO: should we be more explicit here and is this part of the Gateway API contract?
+			// The issue here was that we returned datamodel.ErrWrongKind instead of this resolver error
+			if isErrNotFound(err) {
+				return nil, nil, resolver.ErrNoLink{Name: root, Node: lastPath.Cid()}
+			}
+			return nil, nil, err
+		}
+		lastPath = resolvedSubPath
+		pathRoots = append(pathRoots, lastPath.Cid())
+	}
+
+	pathRoots = pathRoots[:len(pathRoots)-1]
+	return pathRoots, lastPath, nil
+}
+
+// FIXME(@Jorropo): https://github.com/ipld/go-car/issues/315
+type dagStore struct {
+	api *BlocksGateway
+	ctx context.Context
+}
+
+func (ds dagStore) Get(_ context.Context, c cid.Cid) (blocks.Block, error) {
+	return ds.api.blockService.GetBlock(ds.ctx, c)
+}
+
+func (api *BlocksGateway) ResolveMutable(ctx context.Context, p ifacepath.Path) (ImmutablePath, error) {
+	err := p.IsValid()
+	if err != nil {
+		return ImmutablePath{}, err
+	}
+
+	ipath := ipfspath.Path(p.String())
+	switch ipath.Segments()[0] {
+	case "ipns":
+		ipath, err = resolve.ResolveIPNS(ctx, api.namesys, ipath)
+		if err != nil {
+			return ImmutablePath{}, err
+		}
+		imPath, err := NewImmutablePath(ifacepath.New(ipath.String()))
+		if err != nil {
+			return ImmutablePath{}, err
+		}
+		return imPath, nil
+	case "ipfs":
+		imPath, err := NewImmutablePath(ifacepath.New(ipath.String()))
+		if err != nil {
+			return ImmutablePath{}, err
+		}
+		return imPath, nil
+	default:
+		return ImmutablePath{}, fmt.Errorf("unsupported path namespace: %s", p.Namespace())
```

Comment:
nit: in cases like this we  should return explicit http.StatusNotImplemented
```suggestion
		return ImmutablePath{}, NewErrorResponse(fmt.Errorf("unsupported path namespace: %s", p.Namespace()), http.StatusNotImplemented)
```